### PR TITLE
Refactor range from `Attributed` to `Node`s

### PR DESCRIPTION
**Pull request by MichaReiser**
Updates RustPython to https://github.com/RustPython/Parser/pull/22

The main changes are:

* `Attributed` has been removed
* The source code range is now stored on the node directly 
* New `Identifier` and `Constant` types have been introduced (not part of https://github.com/RustPython/Parser/pull/22 but have been changed upstream)
* All node types now have source information attached. 

## Motivation

Adding the range to the node directly will ease passing nodes as a whole to rules rather than destructuring all fields and passing them individually (this was possible before but required using `Located<&ast::IfStmt>` which is a bit more cumbersome to write). 

## Disadvantage

Having the `range` on the nodes has one disadvantage: It now shows up in all match patterns and constructing a new node e.g. for `unparse` requires specifying a range. I think that's an okay compromise.

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #4422** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff/4422" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/charliermarsh/ruff/4422?utm_source=stack-comment).

File: crates/ruff/src/autofix/actions.rs [hunk 1]
```diff
@@ -51,28 +51,28 @@ fn is_lone_child(child: &Stmt, parent: &Stmt, deleted: &[&Stmt]) -> Result<bool>
...
             finalbody,
+            range: _,
         })
-        | StmtKind::TryStar(ast::StmtTryStar {
+        | Stmt::TryStar(ast::StmtTryStar {
             body,
             handlers,
             orelse,
             finalbody,
+            range: _,
```
Comment (charliermarsh) [inline crates/ruff/src/autofix/actions.rs:66]: I tend to bias towards `..`, do you prefer targeted ignores for a specific reason?

Comment (MichaReiser) [reply crates/ruff/src/autofix/actions.rs:66]: It depends. Targeted ignores are useful in code paths where we want to handle all fields (and get errors, if a new field has been added). Now, reading all the code to know if we should handle all fields was too much work, so I opted for targeted ignores (safer).

**Comment by MichaReiser** [comment]
@youknowone any idea why it is now pulling in `ruff_source_location`?

**Comment by youknowone** [comment]
No idea what's happened. I tried clone of this PR and Parser one, and source_location had gone!

**Comment by youknowone** [comment]
> New Identifier and Constant types have been introduced 
Nothing important but just for information. Identifier and Int are added. Constant internal type is also changed to using Idenfier and Int, but Constant itself already existed.

File: crates/ruff/src/checkers/ast/mod.rs [hunk 2]
```diff
@@ -3551,9 +3597,15 @@ where
...
+            Expr::Lambda(
+                lambda @ ast::ExprLambda {
+                    args,
+                    body,
+                    range: _,
+                },
+            ) => {
                 if self.settings.rules.enabled(Rule::ReimplementedListBuiltin) {
-                    flake8_pie::rules::reimplemented_list_builtin(self, expr);
+                    flake8_pie::rules::reimplemented_list_builtin(self, lambda);
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/ruff/src/checkers/ast/mod.rs:3608]: Yeah these are great.

**Comment by MichaReiser** [comment]
> as long as it compiles + tests pass, then it's done well :)

I hope that's true :D. Trust in Rust

**Comment by charliermarsh** [comment]
Trust in Rust

**Comment by MichaReiser** [comment]
> Oh, this is not done yet. #4442 

Unfortunately not. It was a busy afternoon. I plan to merge this pr tomorrow

**Pull request merged**